Add command line arguments to print licenses, authors and donors

Calinou opened the pull request:
This is required for complying with third-party licenses when distributing headless/server binaries (including Docker containers).

Since the license text is already included within the binary, this doesn't increase the binary size.

This adds the following command line arguments:

- `--about license` - Prints Godot's MIT license
- `--about thirdparty-copyrights` - Prints third-party copyright notices with SPDX license names
- `--about thirdparty-licenses` - Prints the full text of third-party licenses
- `--about authors` - Print the engine contributors' names
- `--about donors` - Print the engine donors' names

## Preview

### `--about license`

<img width="799" height="577" alt="image" src="https://github.com/user-attachments/assets/1259109a-282c-4c7e-a9da-a7bcc7af8ea6" />

### `--about thirdparty-copyrights`

<img width="760" height="362" alt="image" src="https://github.com/user-attachments/assets/e08fa3a9-0cea-44d6-891c-44ed18a9a22f" />

### `--about thirdparty-licenses`

<img width="692" height="263" alt="image" src="https://github.com/user-attachments/assets/f5559f99-13b8-420d-bcff-62ddfcbd379e" />

### `--about authors`

<img width="686" height="388" alt="image" src="https://github.com/user-attachments/assets/cbb88a0a-bdf7-4885-a8d8-d419b005c64e" />

### `--about donors`

<img width="1231" height="315" alt="image" src="https://github.com/user-attachments/assets/abc233d4-2491-43db-b16a-0969c6ccf764" />

Calinou committed:
Add command line arguments to print licenses, authors and donors

This is required for complying with third-party licenses when
distributing headless/server binaries (including Docker containers).

Since the license text is already included within the binary,
this doesn't increase the binary size.

bruvzg commented:
> Fix --about authors and --about donors misbehaving on names containing Unicode characters. I can never figure out how to use String::utf8() on a Variant 🙁

```diff
diff --git a/core/config/engine.cpp b/core/config/engine.cpp
index 0b5b5627af..fc304c0631 100644
--- a/core/config/engine.cpp
+++ b/core/config/engine.cpp
@@ -111,7 +111,7 @@ Dictionary Engine::get_version_info() const {
 static Array array_from_info(const char *const *info_list) {
 	Array arr;
 	for (int i = 0; info_list[i] != nullptr; i++) {
-		arr.push_back(info_list[i]);
+		arr.push_back(String::utf8(info_list[i]));
 	}
 	return arr;
 }
@@ -119,7 +119,7 @@ static Array array_from_info(const char *const *info_list) {
 static Array array_from_info_count(const char *const *info_list, int info_count) {
 	Array arr;
 	for (int i = 0; i < info_count; i++) {
-		arr.push_back(info_list[i]);
+		arr.push_back(String::utf8(info_list[i]));
 	}
 	return arr;
 }
```

reduz commented on main/main.cpp:
```diff
@@ -640,6 +642,130 @@ Error Main::setup(const char *execpath, int argc, char *argv[], bool p_second_ph
 			print_line(get_full_version_string());
 			goto error;
 
+		} else if (I->get() == "--about") {
```
I wonder if we could move this to another file like `main/show_license.h/cpp` or similar to avoid adding even more bloat to that file.

Calinou replied:
I'll look into this. Should this be done with some static methods?

Rindbee replied:
Can terminal-related io methods (help information output and command line parameter parsing) be split into a single file?

Calinou replied:
The current CLI parsing implementation needs to be refactored eventually, so it's best done in a separate PR like https://github.com/godotengine/godot/pull/49362 attempted.

Calinou commented:
Rebased and tested again, it works as expected.

Regarding splitting to a separate file, I'm not sure if there are enough lines of code to warrant this. There would be a lot of boilerplate compared to actual code.

nathanfranke commented:
This should comply for desktop, android, HTML5, but not iOS nor UWP [docs](https://docs.godotengine.org/en/stable/about/complying_with_licenses.html#output-log).

Also I would argue a message on startup like "use flag --about for licensing information" is necessary, assuming it isn't already added.

Calinou commented:
> Also I would argue a message on startup like "use flag --about for licensing information" is necessary, assuming it isn't already added.

This could be added, but I've seen complaints about the startup banner already being intrusive.

fire reviewed (approved):
I would like this for 4.3 but @akien-mga let us know if it is a blocker to move the change to a separate file.

fire commented:
Question is there a current feature to add licenses from godot extensions?

Calinou commented:
> Question is there a current feature to add licenses from godot extensions?

Not that I know of, but this should be discussed in its own proposal.

Rindbee commented on main/main.cpp:
```diff
@@ -1097,6 +1099,138 @@ Error Main::setup(const char *execpath, int argc, char *argv[], bool p_second_ph
 			exit_err = ERR_HELP; // Hack to force an early exit in `main()` with a success code.
 			goto error;
 
+		} else if (arg == "--about") {
```
```suggestion
		} else if (arg == "--about-engine") {
```
See above.

Rindbee commented on main/main.cpp:
```diff
@@ -1097,6 +1099,138 @@ Error Main::setup(const char *execpath, int argc, char *argv[], bool p_second_ph
+				if (N->get() == "license") {
+					print_line_rich("[b][u]MIT license[/u][/b]");
+					// Replace reference to the `AUTHORS.md` file by the equivalent command line argument.
+					print_line_rich(String::utf8(GODOT_LICENSE_TEXT).replace("AUTHORS.md", "[code]--about authors[/code] command line argument"));
```
```suggestion
					print_line_rich(String::utf8(GODOT_LICENSE_TEXT).replace("AUTHORS.md", "[code]--about-engine authors[/code] command line argument"));
```
See above.

Rindbee reviewed (approved):
Overall pretty good.

Calinou commented on main/main.cpp:
```diff
@@ -539,6 +540,7 @@ void Main::print_help(const char *p_binary) {
 	print_help_title("General options");
 	print_help_option("-h, --help", "Display this help message.\n");
 	print_help_option("--version", "Display the version string.\n");
+	print_help_option("--about", "Display engine information (\"license\", \"thirdparty-copyrights\", \"thirdparty-licenses\", \"authors\", \"donors\").\n");
```
On the other hand, we have `--help`/`--version` and not `--help-engine`/`--version-engine`. While there's some ambiguity, I think we should favor consistency here.

Rindbee replied:
This is fine with me. I just worry that sensitive game developers will mind game users misunderstanding (mainly the `--about license` option) and protest against this behavior.

Calinou replied:
I've added headings to all printed text to make it more obvious that it refers to the engine license/authors/donors. I think this should suffice.

Calinou commented:
Rebased and tested again, it works as expected. I've added headings to all printed text to make it more obvious that it refers to the engine license/authors/donors.

See OP for updated screenshots.
